syntax = "proto3";

package pyspark.sql.streaming;

message StateRequest {
  int32 version = 1;
  oneof method {
    StatefulProcessorHandleCall statefulProcessorHandleCall = 2;
    ListStateCall listStateCall = 3;
  }
}

message StateResponse {
  int32 statusCode = 1;
  string errorMessage = 2;
}

message StatefulProcessorHandleCall {
  oneof method {
    SetHandleState setHandleState = 2;
    GetListState getListState = 3;
  }
}

message GetListState {
  string stateName = 1;
  string schema = 2;
}

message ListStateCall {
  oneof method {
    Exists exists = 1;
    Get get = 2;
    Update update = 3;
  }
}

message Exists {}

message Get {}

message Update {}

enum HandleState {
  CREATED = 0;
  INITIALIZED = 1;
  DATA_PROCESSED = 2;
  CLOSED = 3;
}

message SetHandleState {
  HandleState state = 1;
}